*: Properly support partial UNIQUE WITHOUT INDEX referencing type descs

Xiang-Gu opened the pull request:
Previously if a partial UNIQUE WITHOUT INDEX references a type descriptor in its predicate, we didn't add back-references in the type descriptor, both in the legacy and declarative schema changer. This commit fixes this.

Fixes #96678
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96681)
<!-- Reviewable:end -->

Xiang-Gu committed:
*: Properly support partial UNIQUE WITHOUT INDEX referencing type descs

Previously if a partial UNIQUE WITHOUT INDEX references a type
descriptor in its predicate, we didn't add back-references in the
type descriptor, both in the legacy and declarative schema changer.
This commit fixes this.

Release note: None

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod the testing request

> Reviewed 3 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96681#-:-NNcyL6iDy1sPyO6EbA-:bv02qub)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Xiang-Gu)

___
*[`pkg/sql/unique_without_index_test.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96681#-NNcyTY6-pZrHtSvOA7k:-NNcyTY6-pZrHtSvOA7l:b-d2yv80) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5b662eec6cf0fcba87602ab5f99113ef7a74766/pkg/sql/unique_without_index_test.go#L53)):*
> ```Go
> 		tdb.Exec(t, "CREATE TYPE typ AS ENUM ('a', 'b');")
> 		tdb.Exec(t, "CREATE TABLE t (i INT PRIMARY KEY, j STRING);")
> 		tdb.Exec(t, "ALTER TABLE t ADD UNIQUE WITHOUT INDEX (j) WHERE (j::typ != 'a');")
> ```

Can you ensure that dropping the enum member fails because it's referenced by the constraint? Then ensure that it succeeds after you drop the constraint?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Should this have a release note?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96681#-:-NNcyhnKA0Z5S60VL9cu:bgjo1tj)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Xiang-Gu)



<!-- Sent from Reviewable.io -->

Xiang-Gu reviewed (commented):
I intentionally didn't put one because this feature UNIQUE WITHOUT INDEX is hidden from users (both in code and in doc).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96681#-:-NNgVok_4vYAgqLeH8ST:busue26)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

Xiang-Gu commented:
TFTR!

bors r+
